stream.go

```
@@ -140,6 +145,203 @@ func (r *Client) StreamPrediction(ctx context.Context, prediction *Prediction) (
+var _ streaming.File = &httpURL{}
+
+func (h *httpURL) Body(ctx context.Context) (io.ReadCloser, error) {
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, h.url, nil)
+	if err != nil {
+		return nil, err
+	}
+	resp, err := h.c.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	return resp.Body, nil
+}
+
+type errWrapper struct {
+	err error
+}
+
+var _ streaming.File = &errWrapper{}
+
+func fileError(err error) streaming.File {
+	return &errWrapper{err: err}
+}
+
+func (e *errWrapper) Body(_ context.Context) (io.ReadCloser, error) {
+	return nil, e.err
+}
+
+func (e *errWrapper) Close() error {
+	return nil
+}
+
+func (r *Client) streamFilesTo(ctx context.Context, out chan<- streaming.File, url string, lastEventID string) {
+	defer close(out)
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	ch := make(chan event)
+	go r.streamEventsTo(ctx, ch, url, lastEventID)
+
+	for e := range ch {
+		url := strings.TrimSuffix(e.rawData, "\n")
+		switch {
+		case strings.HasPrefix(url, "data:"):
+			select {
+			case <-ctx.Done():
+			case out <- &dataURL{url: url}:
+			}
+		case strings.HasPrefix(url, "http"):
+			select {
+			case <-ctx.Done():
+			case out <- &httpURL{c: r.c, url: url}:
+			}
+		default:
+			select {
+			case <-ctx.Done():
+			case out <- fileError(fmt.Errorf("Could not parse URL: %s", url)):
+			}
+			return
+		}
+	}
+}
+
+type event struct {
+	rawData string
+	err     error
+}
+
+func (r *Client) streamEventsTo(ctx context.Context, out chan<- event, url string, lastEventID string) {
+	defer close(out)
+ATTEMPT:
+	for attempt := 0; attempt <= r.options.retryPolicy.maxRetries; attempt++ {
+		req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+		if err != nil {
+			out <- event{err: err}
+			return
+		}
+		req.Header.Set("Accept", "text/event-stream")
+
+		if lastEventID != "" {
+			req.Header.Set("Last-Event-ID", lastEventID)
+		}
+
+		resp, err := r.c.Do(req)
+		if err != nil {
+			select {
+			case <-ctx.Done():
```

Is it ok that we're ignoring any errors from the context throughout this code?